activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb

```diff
@@ -507,6 +507,7 @@ def checkout(checkout_timeout = @checkout_timeout)
       # +conn+: an AbstractAdapter object, which was obtained by earlier by
       # calling #checkout on this pool.
       def checkin(conn)
+        conn.assert_connection_reuseable
```

Comment:
Connections may be checked in (closed) in a dirty state. We should rollback on checkin to close dangling open transactions.